stan/math/prim/err/check_cholesky_factor.hpp

```diff
@@ -25,17 +28,41 @@ namespace math {
  *   factor, if number of rows is less than the number of columns,
  *   if there are 0 columns, or if any element in matrix is NaN
  */
-template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
+template <typename Mat, require_matrix_t<Mat>* = nullptr>
 inline void check_cholesky_factor(const char* function, const char* name,
-                                  const EigMat& y) {
+                                  const Mat& y) {
   check_less_or_equal(function, "columns and rows of Cholesky factor", y.cols(),
                       y.rows());
   check_positive(function, "columns of Cholesky factor", y.cols());
-  const Eigen::Ref<const plain_type_t<EigMat>>& y_ref = y;
+  auto&& y_ref = to_ref(value_of_rec(y));
   check_lower_triangular(function, name, y_ref);
   check_positive(function, name, y_ref.diagonal());
 }
 
+/**
+ * Check if the specified matrix is a valid Cholesky factor.
+ * A Cholesky factor is a lower triangular matrix whose diagonal
+ * elements are all positive.  Note that Cholesky factors need not
+ * be square, but require at least as many rows M as columns N
+ * (i.e., M &gt;= N).
+ * @tparam StdVec A standard vector with inner type inheriting from `MatrixBase`
+ * with dynamic rows and columns.
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y Standard vector of matrices to test
+ * @throw <code>std::domain_error</code> if y is not a valid Cholesky
+ *   factor, if number of rows is less than the number of columns,
+ *   if there are 0 columns, or if any element in matrix is NaN
+ */
+template <typename StdVec, require_std_vector_t<StdVec>* = nullptr>
+void check_cholesky_factor(const char* function, const char* name,
+                           const StdVec& y) {
+  for (size_t i = 0; i < y.size(); ++i) {
+    check_cholesky_factor(function, internal::make_iter_name(name, i).c_str(),
```

Comment:
I'm fine with running a little performance check here, I def expect it to be slower for small matrices though hopefully not much.

> Is there a way to be lazy and avoid creating the name until the check fails?

I tried thinking about this but the only thing I could figure out is to change all of the checks to take in a lambda instead of a const char* that doesn't evaluate until a throw occurs.